enhancements/kube-apiserver/stability-point-to-point-network-check.md

```diff
@@ -0,0 +1,418 @@
+specific `PodNetworkConnectivityCheck` resources.
+
+### 5. Create a binary that can take multiple destinations paired with instances to write to.
+
+### 6. Use that binary in containers included in each kas and oas (or other interesting) pod.
+
+The binary will:
+
+1. List all `PodNetworkConnectivityCheck` resource in the same namespace, where `spec.sourcePodName` and, if specified, 
+   additional label selectors specified in `spec.SourcePodSelector` match is the pod the binary is running in. 
+
+2. Perform each check found.
+
+3. Append entries to the `status.log` and update `status.conditions` as needed.
+
+4. Repeat evey minute.
+
+### User Stories [optional]
+
+Detail the things that people will be able to do if this is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+
+#### Story 1
+
+#### Story 2
+
+### Implementation Details/Notes/Constraints [optional]
+
+What are the caveats to the implementation? What are some important details that
+didn't come across above. Go in to as much detail as necessary here. This might
+be a good place to talk about core concepts and how they relate.
+
+#### RelatedEndpoint Custom Resource
+
+```go
+package v1alpha1
+
+// group: network.openshift.io
+
+import (
+	"time"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// PodNetworkConnectivityCheck
+type PodNetworkConnectivityCheck struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata"`
+
+	Spec   PodNetworkConnectivityCheckSpec   `json:"spec"`
+	Status PodNetworkConnectivityCheckStatus `json:"status"`
+}
+
+type PodNetworkConnectivityCheckSpec struct {
+	// SourcePod names the pod from which the condition will be checked
+	SourcePod string `json:"sourcePod"`
+	// EndpointAddress to check. A TCP address of the form host:port. Note that
+	// if host is a DNS name, then the check would fail if the DNS name cannot
+	// be resolved. Specify an IP address for host to bypass DNS name lookup.
+	TargetEndpoint string `json:"targetEndpoint"`
+}
+
+type PodNetworkConnectivityCheckStatus struct {
+	// Log of check events
+	Log []LogEntry `json:"log"`
+	// Conditions summarize the status of the check
+	Conditions []metav1.ConditionStatus `json:"conditions"`
+}
+
+// LogEntry records events
+type LogEntry struct {
+	// Time of occurrence of check action.
+	Time metav1.Time `json:"time,omitempty"`
+	// Status indicates if the action performed by the check was successful or not.
+	Success bool `json:"status"`
+	// Reason for status in a machine readable format.
+	Reason LogEntryReason `json:"reason"`
+	// Message explaining status in a human readable format.
+	Message string `json:"message"`
+	// Latency records how long the action mentioned in the entry took.
+	Latency time.Duration `json:"latency"`
```

Comment:
This is a metric.  Why are you reporting it as an API request

Reply:
> This is a metric. Why are you reporting it as an API object

our troubles don't happen independently of when monitoring is up.  When access is lost, we need to be able to write this data and a localhost write is much more reliable than a metric scrape.  Especially since metrics are pulls and all metrics that aren't pulled during a rollout are fully lost (see our previous: metrics are missing for rollouts, *again* problems). The API has prove much more reliable.